llama_index/llm_predictor/huggingface.py

```diff
@@ -0,0 +1,229 @@
+
+        class StopOnTokens(StoppingCriteria):
+            def __call__(
+                self,
+                input_ids: torch.LongTensor,
+                scores: torch.FloatTensor,
+                **kwargs: Any,
+            ) -> bool:
+                for stop_id in stopping_ids_list:
+                    if input_ids[0][-1] == stop_id:
+                        return True
+                return False
+
+        self._stopping_criteria = StoppingCriteriaList([StopOnTokens()])
+
+    def get_llm_metadata(self) -> LLMMetadata:
+        """Get LLM metadata."""
+        return LLMMetadata(
+            max_input_size=self._max_input_size, num_output=self._max_new_tokens
+        )
+
+    def stream(self, prompt: Prompt, **prompt_args: Any) -> Tuple[Generator, str]:
+        """Stream the answer to a query.
+
+        NOTE: this is a beta feature. Will try to build or use
+        better abstractions about response handling.
+
+        Args:
+            prompt (Prompt): Prompt to use for prediction.
+
+        Returns:
+            str: The predicted answer.
+
+        """
+        import torch
+        from transformers import TextIteratorStreamer
+
+        formatted_prompt = prompt.format(**prompt_args)
+        full_prompt = self._query_wrapper_prompt.format(query_str=formatted_prompt)
+        if self._system_prompt:
+            full_prompt = f"{self._system_prompt} {full_prompt}"
+
+        inputs = self.tokenizer(full_prompt, return_tensors="pt")
+        if "cuda" == self._device_map or (
+            "auto" == self._device_map and torch.cuda.is_available()
+        ):
+            inputs = inputs.to("cuda")
+
+        streamer = TextIteratorStreamer(
+            self.tokenizer,
+            skip_prompt=True,
+            decode_kwargs={"skip_special_tokens": True},
+        )
+        generation_kwargs = dict(
+            inputs,
+            streamer=streamer,
+            max_new_tokens=self._max_new_tokens,
+            temperature=self._temperature,
+            do_sample=self._do_sample,
+            stopping_criteria=self._stopping_criteria,
+        )
+
+        # generate in background thread
+        # NOTE/TODO: token counting doesn't work with streaming
+        thread = Thread(target=self.model.generate, kwargs=generation_kwargs)
+        thread.start()
+
+        # create generator based off of streamer
+        def response() -> Generator:
+            for x in streamer:
+                yield x
+
+        return response(), formatted_prompt
+
+    @property
+    def total_tokens_used(self) -> int:
+        """Get the total tokens used so far."""
+        return self._total_tokens_used
+
+    @property
+    def last_token_usage(self) -> int:
+        """Get the last token usage."""
+        return self._last_token_usage or 0
+
+    @last_token_usage.setter
+    def last_token_usage(self, value: int) -> None:
+        """Set the last token usage."""
+        self._last_token_usage = value
```

Comment:
not for this PR: we should refactor this away. LLMPredictor should not hold this state.